django/db/models/fields/json.py

```diff
@@ -390,6 +394,19 @@ def as_sqlite(self, compiler, connection):
             "THEN JSON_TYPE(%s, %%s) ELSE JSON_EXTRACT(%s, %%s) END)"
         ) % (lhs, datatype_values, lhs, lhs), (tuple(params) + (json_path,)) * 3
 
+    def get_update_expression(self, value, lhs=None):
+        from ..functions import JSONRemove, JSONSet
+
+        field, key_transforms = self.unwrap_transforms()
+
+        if lhs is None:
+            lhs = field
+
+        if value is NOT_PROVIDED:
```

Comment:
I like the idea of using a sentinel to differentiate between `null` and key deletions ([it's something that I believe is missing](https://code.djangoproject.com/ticket/35381#comment:3) for JSON `null` vs  SQL `NULL`) bet I'm not sure we should piggy back on that one since it was meant for a completely different purpose.

Reply:
I agree @charettes! This was a temporary solution as suggested by Markus the other day, and I linked to your exact comment: https://fosstodon.org/@laymonage/112949402890177435 (Like I said, big fan 😂)

> I think the idea of making `for_save` a `(enum.Enum, str)` to differentiate between different scenario could could as most usage out there are doing `if for_save` and most likely don't need to differentite between both.

That's what I thought too – though I suppose it could still break if people check for the boolean explicitly e.g. `if for_save is True`. Still doable, just need to be more careful when we do the transition.

> My immediate concerns with these changes are that they are possibly trying to do too much at once?
> 
> Is there a reason why we can't first land support for `JSONSet` / `JSONRemove` in a first set of changes targeted for 5.2, let them simmer for a release, and then consider adding support for the transform combining feature later on?

No particular reason, really. When we implemented the first pass for the `JSONSet` and `JSONRemove` functions, we were curious how much work it would take to implement the transforms support in `update()` directly. Turns out it wasn't much, but indeed it's a bit more "controversial".

Sarah and I discussed this at the DjangoCon US sprints and we decided to split this PR into two. One for the DB functions, and one for the nicer syntactic sugar via `update()` directly – just like you said. I noted this in my comment https://github.com/django/django/pull/18489#pullrequestreview-2395014247 just yesterday.

@adzshaf and I are still working on polishing the DB functions to be ready for a new, smaller PR. We found a few more edge cases so it's taking us a bit longer than we thought. We'll post an update on the ticket when we put the PR up 😄 

> I like the ideas put forward by this patch and the separation of concern between field and transform logic but these are more fundamental changes to the ORM works and bring larger questions. For example, should we support `save(update_fields)` with transforms as well? Is it time to introduce a specialized type of transform that denotes nested data structure access instead of trying to push the current abstraction in this space?
> 
> IMO these questions can likely be more easily answered while users are allowed to play with `JSONSet` and `JSONRemove` and these primitives are iterated on to iron out bugs first.

Thank you, and yes there are definitely more areas of the ORM that we haven't considered like `bulk_{create,update}`, `update_or_create`, etc. and `save(update_fields)` is one of them. I'd be happy to focus on getting `JSONSet` and `JSONRemove` in first, then discuss the more uncertain bits after.

Reply:
Thanks for the update, it's great to see you push all of these JSONField improvement forward. It's has been a game changer to be able to use `JSONField` for non-Postgres database and a lot of that is thanks to you 🙇 

I think that the two-commit approach has a lot of merit considering how hard it is to get these things right the first time. Let me know when things have shappen if you'd like more feedback. As I said already I think the proposed approach is great in the sense that it demonstrates that it can be done without too many invasive changes.